lints/duplicate_mutable_accounts/src/lib.rs

```diff
@@ -0,0 +1,248 @@
+    /// Lists of Anchor accounts found in structs that derive Anchor `Accounts` trait, partitioned by Anchor account type
+    anchor_accounts: HashMap<DefId, Vec<(Symbol, Span)>>,
+    /// List of Anchor `#[account]` macro  constraints
+    anchor_macro_constraints: Streams,
+    /// List of pairs of Anchor accounts with same types, without any alternate constraint
+    spans: Vec<(Span, Span)>,
+    /// Indicates if alternate constraints were used or not
+    no_alternate_constraints: bool,
+}
+
+impl<'tcx> LateLintPass<'tcx> for DuplicateMutableAccounts {
+    fn check_struct_def(&mut self, cx: &LateContext<'tcx>, variant_data: &'tcx VariantData<'tcx>) {
+        if let VariantData::Struct(fields, _) = variant_data {
+            fields.iter().for_each(|field| {
+                if_chain! {
+                    if let Some(def_id) = get_def_id(field.ty);
+                    let middle_ty = cx.tcx.type_of(def_id);
+                    if match_type(cx, middle_ty, &paths::ANCHOR_LANG_ACCOUNT);
+                    if let Some(account_id) = get_anchor_account_type_def_id(field);
+                    then {
+                        if let Some(v) = self.anchor_accounts.get_mut(&account_id) {
+                            v.push((field.ident.name, field.span));
+                        } else {
+                            self.anchor_accounts
+                                .insert(account_id, vec![(field.ident.name, field.span)]);
+                        }
+                    }
+                }
+            });
+        }
+    }
+
+    fn check_attribute(&mut self, _: &LateContext<'tcx>, attribute: &'tcx Attribute) {
+        if_chain! {
+            if let AttrKind::Normal(attr_item, _) = &attribute.kind;
+            let name = attribute.name_or_empty();
+            if name.as_str() == "account";
+            if let MacArgs::Delimited(_, _, token_stream) = &attr_item.args;
+            then {
+                self.anchor_macro_constraints.0.push(token_stream.clone());
+            }
+        }
+    }
+
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        _: FnKind<'tcx>,
+        _: &'tcx FnDecl<'tcx>,
+        body: &'tcx Body<'tcx>,
+        span: Span,
+        _: HirId,
+    ) {
+        if !span.from_expansion() {
+            let mut values = Values::new(cx);
+            values.get_referenced_accounts_and_if_statements(body);
+
+            values.accounts.values().for_each(|exprs| {
+                if exprs.len() > 1 {
+                    self.no_alternate_constraints = true; // assume no alternate constraints
+                    for current in 0..exprs.len() - 1 {
+                        for next in current + 1..exprs.len() {
+                            if values.check_key_constraint(exprs[current], exprs[next]) {
+                                // if there is at least one alt constraint, set flag to false
+                                self.no_alternate_constraints = false;
```

Comment:
This could flip back to `true` if the program contains more than one function, right? (Or am I misunderstanding?)

I'm wondering if `no_alternate_constraints` and `spans` should be maintained per function, similar to how `anchor_accounts` is maintained per struct def.